prometheus/main.go

```diff
@@ -0,0 +1,380 @@
+	a.HelpFlag.Short('h')
+
+	cfg := &config{
+		awsRegion:        os.Getenv("REGION"),
+		awsEnv:           os.Getenv("ENV"),
+		influxdbPassword: os.Getenv("INFLUXDB_PW"),
+		promlogConfig:    promlog.Config{},
+	}
+
+	a.Flag("atny-url", "The URL of the remote Aternity APM CMX endpoint to send samples to. None, if empty.").
+		Default("").StringVar(&cfg.aternityURL)
+	a.Flag("atny-cmx-dimensions", "A csv list of dimensions and values. None, if empty.").
+		Default("").StringVar(&cfg.aternityCmxDimensions)
+	a.Flag("graphite-address", "The host:port of the Graphite server to send samples to. None, if empty.").
+		Default("").StringVar(&cfg.graphiteAddress)
+	a.Flag("graphite-transport", "Transport protocol to use to communicate with Graphite. 'tcp', if empty.").
+		Default("tcp").StringVar(&cfg.graphiteTransport)
+	a.Flag("graphite-prefix", "The prefix to prepend to all metrics exported to Graphite. None, if empty.").
+		Default("").StringVar(&cfg.graphitePrefix)
+	a.Flag("opentsdb-url", "The URL of the remote OpenTSDB server to send samples to. None, if empty.").
+		Default("").StringVar(&cfg.opentsdbURL)
+	a.Flag("influxdb-url", "The URL of the remote InfluxDB server to send samples to. None, if empty.").
+		Default("").StringVar(&cfg.influxdbURL)
+	a.Flag("influxdb.retention-policy", "The InfluxDB retention policy to use.").
+		Default("autogen").StringVar(&cfg.influxdbRetentionPolicy)
+	a.Flag("influxdb.username", "The username to use when sending samples to InfluxDB. The corresponding password must be provided via the INFLUXDB_PW environment variable.").
+		Default("").StringVar(&cfg.influxdbUsername)
+	a.Flag("influxdb.database", "The name of the database to use for storing samples in InfluxDB.").
+		Default("prometheus").StringVar(&cfg.influxdbDatabase)
+	a.Flag("send-timeout", "The timeout to use when sending samples to the remote storage.").
+		Default("30s").DurationVar(&cfg.remoteTimeout)
+	a.Flag("web.listen-address", "Address to listen on for web endpoints.").
+		Default(":9201").StringVar(&cfg.listenAddr)
+	a.Flag("web.telemetry-path", "Address to listen on for web endpoints.").
+		Default("/metrics").StringVar(&cfg.telemetryPath)
+
+	flag.AddFlags(a, &cfg.promlogConfig)
+
+	_, err := a.Parse(os.Args[1:])
+	if err != nil {
+		fmt.Fprintln(os.Stderr, errors.Wrapf(err, "Error parsing commandline arguments"))
+		a.Usage(os.Args[1:])
+		os.Exit(2)
+	}
+
+	return cfg
+}
+
+type writer interface {
+	Write(samples model.Samples) error
+	Name() string
+}
+
+type reader interface {
+	Read(req *prompb.ReadRequest) (*prompb.ReadResponse, error)
+	Name() string
+}
+
+func buildClients(logger log.Logger, cfg *config) ([]writer, []reader) {
+	var writers []writer
+	var readers []reader
+	if cfg.graphiteAddress != "" {
+		c := graphite.NewClient(
+			log.With(logger, "storage", "Graphite"),
+			cfg.graphiteAddress, cfg.graphiteTransport,
+			cfg.remoteTimeout, cfg.graphitePrefix)
+		writers = append(writers, c)
+	}
+	if cfg.opentsdbURL != "" {
```

Comment:
do we need to handle these other clients? or is this only an adapter for CMX? I am not sure how the model works.

Reply:
Addressed.